Header Menu and SubMenu Test 118 to 121

Amogh-Hirameki opened the pull request:
```
Feature: To check Header Links

@HeaderValidation
Scenario: Validate Header Links

When User hovers on header Links
Then Validate that links are visible
And Validate that menus are working
And Validate that sub menus are working



```

Amogh-Hirameki committed:
Header Menu and SubMenu Test cases

Header : -
Feature File : Done
Page Object file : Done
Step Definition file : Done

csonparote commented on platform/src/test/java/platformfeatures/PF_Header.feature:
```diff
@@ -0,0 +1,13 @@
+Feature: To check Header Links
+
+@FunctionalTest
```
Please add exclusive tag to scenario

Amogh-Hirameki replied:
Done Changes as requested

csonparote commented on platform/src/test/java/platformfeatures/PF_Header.feature:
```diff
@@ -0,0 +1,13 @@
+@FunctionalTest
+Scenario Outline: Validate Header Links
+
+Given Navigate to URL and open Header <browser>
```
Seperate opening and closing browser into different steps

Amogh-Hirameki replied:
Done Changes as requested

csonparote replied:
Feature file needs more details:
Given open browser
And navigate to url
When -user performs action
And -user performs another action-
Then -validate test result-
And close browser

Please refer online resources for feature file writing in cucumber

Amogh-Hirameki replied:
Done Changes as requested

csonparote commented on platform/src/test/java/platformstepdefinition/PF_Header.java:
```diff
@@ -0,0 +1,24 @@
+@RunWith(Cucumber.class)
+public class PF_Header extends NewBaseClass {
+
+	public WebDriver driver;
```
Replace with public WebDriver driver = ObjectHelper.driver; 
Needed to retain driver instance

Amogh-Hirameki replied:
Done Changes as requested

csonparote commented on platform/src/test/java/platformstepdefinition/PF_Header.java:
```diff
@@ -0,0 +1,24 @@
+	public void navigate_to_url_(String browser) throws Throwable {
+		driver = openbrowser(browser);
+		Header headerobj = new Header(driver);
+		headerobj.headerAndSubMenuLinksVisibilityChecking();
```
Refactor method name to verb-noun format
Ex: checkHeaderAndSubMenuLinksVisibility();

Amogh-Hirameki replied:
Done Changes as requested

csonparote replied:
Please rearrange method name in verb noun format
Ex: headerAndSubMenuLinksVisibilityChecking(); becomes checkHeaderAndSubMenuLinksVisibility()

It is helpful for global team to understand

Amogh-Hirameki replied:
Done Changes as requested

csonparote commented on platform/src/test/java/platformstepdefinition/PF_Header.java:
```diff
@@ -0,0 +1,24 @@
+		driver = openbrowser(browser);
+		Header headerobj = new Header(driver);
+		headerobj.headerAndSubMenuLinksVisibilityChecking();
+		headerobj.headerLinksChecking();
```
Refactor method name to verb-noun format

Amogh-Hirameki replied:
Done Changes as requested

csonparote replied:
Please rearrange method name in verb noun format
Ex: headerAndSubMenuLinksVisibilityChecking(); becomes checkHeaderAndSubMenuLinksVisibility()

It is helpful for global team to understand

Amogh-Hirameki replied:
Done Changes as requested

csonparote replied:
Its Good

csonparote commented on platform/src/test/java/platformstepdefinition/PF_Header.java:
```diff
@@ -0,0 +1,24 @@
+		Header headerobj = new Header(driver);
+		headerobj.headerAndSubMenuLinksVisibilityChecking();
+		headerobj.headerLinksChecking();
+		headerobj.subMenuLinkChecking();
```
Refactor method name to verb-noun format

Amogh-Hirameki replied:
Done Changes as requested

csonparote replied:
Please refactor as per above comment

csonparote commented on platform/src/main/java/platform/pageobjects/Header/Header.java:
```diff
@@ -0,0 +1,362 @@
+			log.info("Close Zkai pop up");
+		}
+
+		Thread.sleep(5000);
```
Replace with wait from CommonFunctions.
Please remove Thread.sleep from code

Amogh-Hirameki replied:
Done Changes as requested

Sopan-Patil commented on platform/src/main/java/platform/pageobjects/Header/Header.java:
```diff
@@ -0,0 +1,362 @@
+	@FindBy(xpath = "//a[@class='logo logo-white']")
+	public WebElement stagiaLogo;
+
+	@FindBy(xpath = "//a[@class='menu__link js-not-link'][contains(text(),'受験の窓口')]")
```
Not to use the content. You can use other locators.

Amogh-Hirameki replied:
Done Changes as requested

Sopan-Patil replied:
@FindBy(xpath = "//span[contains(text(),'ログイン')]")
	public WebElement loginButton;
Please do not use content

Sopan-Patil replied:
Please use assert to validate for the link page is opened or not

Sopan-Patil commented on platform/src/main/java/platform/pageobjects/Header/Header.java:
```diff
@@ -0,0 +1,362 @@
+
+	// Xpaths for the Sub Menus
+
+	@FindBy(xpath = "/html/body/div/div[1]/div[1]/div/div[1]/ul/li[1]/div/ul/li[1]/a/span")
```
XPath not to use4d absolute xpath. It will relative xpath.

Amogh-Hirameki replied:
Done Changes as requested

Sopan-Patil commented on platform/src/main/java/platform/pageobjects/Header/Header.java:
```diff
@@ -0,0 +1,362 @@
+
+	@FindBy(xpath = "/html/body/div/div[1]/div[1]/div/div[1]/ul/li[2]/div/ul/li[1]/a/span")
+	public WebElement learningTopLink;
+
```
XPath not to use4d absolute xpath. It will relative xpath.

Amogh-Hirameki replied:
Done Changes as requested

Sopan-Patil commented on platform/src/main/java/platform/pageobjects/Header/Header.java:
```diff
@@ -0,0 +1,362 @@
+
+	public void subMenuLinkChecking() throws Exception {
+
+		Thread.sleep(5000);
```
Please use condition wait.

Amogh-Hirameki replied:
Done Changes as requested

Amogh-Hirameki committed:
Modified the Code

- Added exclusive tag to scenario
- Separated opening and closing browser into different steps
- Replaced with public WebDriver driver = ObjectHelper.driver
- Replaced method name to verb-noun format
- Replaced Thread.sleep with wait from CommonFunctions
- Replaced Absolute XPath with Relative XPath

Amogh-Hirameki committed:
Update Header.java

Amogh-Hirameki committed:
Modified Header

- Removed Remaining Thread.sleep

csonparote commented:
Added comments on my review and resolved conversation for changes made

Amogh-Hirameki committed:
Modified Changes

- Created Detailed Feature File

Sopan-Patil commented:
> Header : -
> Feature File : Done
> Page Object file : Done
> Step Definition file : Done

Please put feature file content here

Sopan-Patil reviewed (changes requested):
test case is not running.

[RemoteTestNG] detected TestNG version 6.13.0
org.testng.TestNGException: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at org.testng.TestNG.parseSuite(TestNG.java:313)
	at org.testng.TestNG.initializeSuitesAndJarFile(TestNG.java:334)
	at org.testng.TestNG.initializeEverything(TestNG.java:974)
	at org.testng.remote.support.RemoteTestNG6_12.initialize(RemoteTestNG6_12.java:22)
	at org.testng.remote.AbstractRemoteTestNG.run(AbstractRemoteTestNG.java:98)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:251)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:77)
Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:145)
	at sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:131)
	at java.security.cert.CertPathBuilder.build(CertPathBuilder.java:280)
	at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:382)
	at sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:292)
	at sun.security.validator.Validator.validate(Validator.java:260)
	at sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:324)
	at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:229)
	at sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:124)
	at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1351)
	at sun.security.ssl.ClientHandshaker.processMessage(ClientHandshaker.java:156)
	at sun.security.ssl.Handshaker.processLoop(Handshaker.java:925)
	at sun.security.ssl.Handshaker.process_record(Handshaker.java:860)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1043)
	at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1343)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1371)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1355)
	at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:563)
	at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:185)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1511)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1439)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:254)
	at org.apache.xerces.impl.XMLEntityManager.setupCurrentEntity(Unknown Source)
	at org.apache.xerces.impl.XMLEntityManager.startEntity(Unknown Source)
	at org.apache.xerces.impl.XMLEntityManager.startDTDEntity(Unknown Source)
	at org.apache.xerces.impl.XMLDTDScannerImpl.setInputSource(Unknown Source)
	at org.apache.xerces.impl.XMLDocumentScannerImpl$DTDDispatcher.dispatch(Unknown Source)
	at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanDocument(Unknown Source)
	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
	at org.apache.xerces.parsers.XMLParser.parse(Unknown Source)
	at org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source)
	at org.apache.xerces.jaxp.SAXParserImpl$JAXPSAXParser.parse(Unknown Source)
	at org.apache.xerces.jaxp.SAXParserImpl.parse(Unknown Source)
	at javax.xml.parsers.SAXParser.parse(SAXParser.java:195)
	at org.testng.xml.XMLParser.parse(XMLParser.java:38)
	at org.testng.xml.SuiteXmlParser.parse(SuiteXmlParser.java:16)
	at org.testng.xml.SuiteXmlParser.parse(SuiteXmlParser.java:9)
	at org.testng.xml.Parser.parse(Parser.java:152)
	at org.testng.xml.Parser.parse(Parser.java:233)
	at org.testng.TestNG.parseSuite(TestNG.java:290)
	... 6 more

csonparote commented on platform/src/main/java/platform/pageobjects/Header/Header.java:
```diff
@@ -0,0 +1,390 @@
+		log.info("Opened Test Top submenu");
+		driver.navigate().back();
+		log.info("Back on Previous Page");
+
```
It seems you are performing same steps in the code from 269-278
Plz write a method and pass links as parameters

Amogh-Hirameki replied:
@csonparote 
Can you please elaborate ?

csonparote replied:
U can also pass an array of key value pair to clickSubMenu() to verify links .
NVM this is good too.
No changes needed.
Resolving conversation.

csonparote reviewed (changes requested):
Plz make changes

Amogh-Hirameki committed:
Merge branch 'main' into Amogh_Header_PT_118_121

Amogh-Hirameki committed:
Modified Changes

- Updated Feature File
- Updated Step Definition File
- Create a Function for similar Steps

csonparote reviewed (commented):
Done PR review

Sopan-Patil commented:
Please work on tab closing and pl up to date your brach with main

Amogh-Hirameki committed:
Merge branch 'main' into Amogh_Header_PT_118_121

Amogh-Hirameki committed:
Merge branch 'main' into Amogh_Header_PT_118_121

Amogh-Hirameki committed:
Merge branch 'main' into Amogh_Header_PT_118_121

Amogh-Hirameki committed:
Modified PageObject file

- Added Code to close the tabs

Amogh-Hirameki committed:
Merge branch 'main' into Amogh_Header_PT_118_121

Sopan-Patil reviewed (changes requested):
Getting error on script when executing

Amogh-Hirameki committed:
Update Header.java

Amogh-Hirameki committed:
Merge branch 'main' into Amogh_Header_PT_118_121
